source/util/FeatureRegistry.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "FeatureRegistry.h"
+#include "../logging/LoggerFactory.h"
+
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    for (auto &feature : features)
+    {
+        if (feature->getName() == name)
+        {
+            return feature;
+        }
+    }
+
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::shared_ptr<Feature> feature)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (std::find(features.begin(), features.end(), feature) != features.end())
+    {
+        LOGM_WARN(TAG, "Attemped to add %s to Feature Registry despite already existing", feature->getName().c_str());
+    }
+    else
+    {
+        features.push_back(feature);
+    }
+}
+
+void FeatureRegistry::remove(const std::string &name)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    for (int i = 0; (unsigned)i < features.size(); i++)
+    {
+        if (features.at(i)->getName() == name)
+        {
+            features.erase(features.begin() + i);
+            break;
+        }
+    }
+}
+
+int FeatureRegistry::getSize()
+{
+    return features.size();
+}
+
+void FeatureRegistry::stopAll()
+{
+    while (!features.empty())
```

Comment:
This logic is strange. Assuming there are no side-effects it will just call `Feature::stop` on the same index in an endless loop.  Based on the name I would expect that this function is symmetric with `startAll` and iterates through all the features and calls `Feature::stop`. If the intent is to both `stop` a feature and do something else (I think that it is), that seems flawed and let's talk about that offline.

Similarly strange is the fact that `startAll` acquires a lock but `stopAll` does not.